README
	        --[  arp-sk Options manipulation  ]--

	      Frederic Raynal <pappy@security-labs.org>



-- --[  Modes  ]-- --

There are 2 basics mode:
  - who-has: build a request ARP message. 
  - reply: build a reply ARP message (default)

Other advanced modes should come very soon
  - arping: send a who-has to every host on the LAN to see who is here
  - promisc: detection of boxes that are sniffing on the network using
             promiscuous mode of their network interface
  - arpmim: perform Man in the Middle attack


-- --[ Link level options  ]-- --

-s: set the source address of the packet. 
    Default : MAC address of the interface used to send the packets.

-d: set the destination address of the packet
    Default: broadcast

These 2 options have a strong influence on the ARP message itself.
Here are the default according to these options:

  - request

      # ./arp-sk -i eth1 -w
      + Running mode "who-has"
      + IfName: eth1
      + Source MAC: 52:54:05:f4:62:30
      + Source ARP MAC: 52:54:05:f4:62:30
      + Source ARP IP : 192.168.1.1 (batman)
      + Target MAC: ff:ff:ff:ff:ff:ff
      + Target ARP MAC: 00:00:00:00:00:00
      + Target ARP IP : 255.255.255.255 (255.255.255.255)


  - reply

      # ./arp-sk -i eth1 -r
      + Running mode "reply"
      + IfName: eth1
      + Source MAC: 52:54:05:f4:62:30
      + Source ARP MAC: 52:54:05:f4:62:30
      + Source ARP IP : 192.168.1.1 (batman)
      + Target MAC: ff:ff:ff:ff:ff:ff
      + Target ARP MAC: ff:ff:ff:ff:ff:ff
      + Target ARP IP : 255.255.255.255 (255.255.255.255)

  The only difference comes from the destiantion mac address from ARP
message, since it has to be 00:00:00:00:00:00. For the reply mode,
consistency is preserved and the destination MAC address used for the
link layer is copied in the ARP message.


-- --[  ARP message  ]-- --

  This default comportment is almost dumb, especially if you look at
the destination part of the ARP message. Hence, it is necessary to
provide destination addresses to avoid that. You can play with -D
which accepts several inputs -D <hostname|hostip>[:MAC]

If you use at the same time -d and -D, everything you specify is put
exactly where it should :
  -d in the link layer
  -D in the ARP message
If you don't provide a MAC address with the -D, the one from the link
layer is used.


The logic is almost the same for the source parts of the ARP message.
If you intend to spoof addresses, you will need to play with the
argument  -S <hostname|hostip>[:MAC] 

Note that several writings are legal :

  - just set the IP address -> -S 1.1.1.1
  - just set the MAC address -> -S :11:11:11:11:11:11
    DO NOT FORGET THE ':' AT THE BEGINNING, otherwise, arp-sk will
    consider there is a mistake here
  - both IP and MAC -> -S 1.1.1.1:11:11:11:11:11:11


Of course, you can perform the same tricks with the destination (-D)


-- --[  A feature, not a bug ...  ]-- --

If you provide a MAC address for -S or -D, but no IP one, the arping
way is used to lookup the IP address. So for instance:

    # ./arp-sk -c 1 -t 1 --beep -i eth1 -s 11:11:11:11:11:11 -D 1.2.3.4 --use-ts
    option beep
    - Warning: no mode given, using default.
    + Running mode "reply"
    + IfName: eth1
    + Source MAC: 11:11:11:11:11:11
 -> Wrote ICMP timestamp (54 bytes) to retrieve IP address ... 
    - Warning: can't resolve from mac 11:11:11:11:11:11 => use local one.
    + Source ARP MAC: 11:11:11:11:11:11
    + Source ARP IP : 192.168.1.1 (batman)
    + Target MAC: ff:ff:ff:ff:ff:ff
    + Target ARP MAC: ff:ff:ff:ff:ff:ff
    + Target ARP IP : 1.2.3.4 (1.2.3.4)

    --- Start sending --
    (1) To: ff:ff:ff:ff:ff:ff From: 11:11:11:11:11:11 0x0806
        ARP For 1.2.3.4 (ff:ff:ff:ff:ff:ff):
            192.168.1.1 is at 11:11:11:11:11:11

    --- 1.2.3.4 (ff:ff:ff:ff:ff:ff) statistic ---
    To: ff:ff:ff:ff:ff:ff From: 11:11:11:11:11:11 0x0806
        ARP For 1.2.3.4 (ff:ff:ff:ff:ff:ff):
            192.168.1.1 is at 11:11:11:11:11:11
    1 packets tramitted (each: 96 bytes - total: 96 bytes)

The -s flag is used to set the mac address at the link layer. Since no
information is provided for the ARP message, arp-sk attempts to
retrieve the IP address bound to 11:11:11:11:11:11. Thus, it sends an
icmp-timestamp to a broadcast address and wait for an answer. 

Since it does not receive any, it choses to use the local IP address,
and arp-sk goes its way.

The --use-ts option forces the sending of the icmp-timestamp
message. The default behavior is NOT to send it and use the local IP
address.

-- --[  Address resolution  ]-- --

There are 2 ways to lookup an address. 

You have the IP and look for the MAC address:
Then, you look in the ARP cache (/proc/net/arp with Linux, sysctl()
for BSD). If you can't find the entry you are looking for, arp-sk
sends a fake udp message (dns) to the target, that answers with an
icmp-unreach ... then you have its MAC address in your ARP cache.

You have the MAC and look for the IP address:
Here, you broadcast a ping with MAC address filled with the target
MAC, but IP address set to broadcast address. To answer, the atrget
needs to send an arp who-has to you, so you have its address.

But ...

1. If the target already has you in its cache, the who-has is not
   send, but you receive the icmp-reply anyway => you don't have the
   IP address.

2. not all systems do answer to broadcast icmp messages. For instance,
   windows never answers by default, and Unix can be configured to
   drop such messages.

Capturing the received answer is a way to bypass the 1st point, but
that would require to use libpcap, which is not (yet) in my plan ;-) 

Anybody having a solution for the second point can drop me a mail.


ChangeLog
                        --[  ChangeLog  ]--

              Frederic Raynal <pappy@security-labs.org>



Patch or bug fix' authors are specified within [XX], where XX are initials
of first name and last name of the author (see CONTRIB files). Those
without any [XX] are mine.



Sun Dec  5 15:55:19 CET 2004                               0.0.16

  [Changes for users]
    - remove the -a (address space) options (was useless)
    - the -t (timing) is now -T (-t will be used for futur compatibility
      with arpspoof)
    - configure: lots of improvements
        + unneeded installs are removed
        + removed --with-arpcache=FILE (will make it system dependent)
        + "make distcheck" is supported
        + test the version of libnet, ...

Fri Oct 11 00:44:29 CEST 2002                               0.0.15

  [General]
    - name of the device is stored in a global (sk_ifname), defined in
      sendpacket.c
	
  [lookup-eth-ip.c]
    - resolve_ip_from_eth(): BUG FIX
	use a specific libnet_t* link to send the icmp message. Using the
	same as the one for arp caused malformed messages to be sent.
    - add native support for BSD lookup of MAC/IP addresses
    - lookup_in_arp_cache(): BUG FIX [LL]
        when the cache contains
          10.0.0.11 ...
          10.0.0.1  ... 
      and that you look for 10.0.0.1, it retrieves 10.0.0.11



Sun Sep 15 15:46:00 CEST 2002                               0.0.14

  [sk.c]
    - use va_list in fatal() and warning()
        
  [General]
    - move src/*.h to include/
    - remove the handler structure to use directly pointers to functions
    - changed my coding convetions making sources hopefully more readable
    - all structures are now called arpsk_*_t
    - add ARP file [CB]
    - add --with-arpcache=FILE in configure.in


Tue Aug 27 11:12:36 CEST 2002                               0.0.13

  [arp-common.c/eth-common.c]
  - Fix a bug that could lead to failure when lots of packets where sent
    using the random hook: the libnet_build_*() functions were using 0 as
    last argument instead of the ptag to rebuild 
        => a new pblock was created each time
        => more and mode pblock where added
        => packet became bigger and bigger until it exceeded the max
           allowed size
        => libnet_write()->libnet_write_link()->sendto() failed
        
  [arp-sk]
  - add -N: force name resolution during display

  [General]
  - remove everything that was related to packet pool
  - complete the man page and add a Makefile to it
  - replace _SIGNAL_SENDING_ by HAVE_SIGSEND
  - add HAVE_SIGSEND by default in the configure.in.
  - change default install directory from /usr/local/bin to
    /usr/local/sbin
        
        
  [sendpacket.c]
  - former sendpackets.c, now, it just send ONE packet to the wire
  - made cleanup in libnet writing functions: simply rests on
    libnet_write() that does all the job
  - add an option to force address resolution (default is now no name
    resolution unless -N is used)
  - add calls to pause() when signals are used for sending, this make the
    program is no more CPU consuming
  - fix a memory leak when leaving arp-sk with cleanup() (forgot to free
    sk_pkt) 
       
Tue Aug  6 16:21:25 CEST 2002                               0.0.12

  [General]
  - add configure (my first so feel free to improve it ;-)
  - add a libcompat
  - compile and run under FreeBSD 4.4
        
  [ip-common.c]
  ip_is_reserved(): change the generation of random IP addresses since
    Fyodor provides a new one in nmap-3.0


  [lookup-eth-ip.c]
  lookup_in_arp_cache(): the former way of checking incomplete addresses
    was buggy. Now, I look at the flag of the entry on the cache (0x0 means
    incomplete). 

Wed Jul 17 12:01:23 CEST 2002                               0.0.11b

  [eth-common.c]
  - change a structure in get_if_eth_addr(), conforming to the new return
    type of libnet_get_hwaddr() that came with libnet-1.1.0-beta7b

  [Makefile]
  - clean up and add the creation of the directory $(DO)/Objects

  [README]
  - add information in README about address resolution and options
    manipulation 

  [INSTALL]
  - create INSTALL file


Sun Jun 30 15:44:51 CEST 2002                               0.0.11

  [arp-sk]
  - add support for --rand* (support while arguments are handled need to
    be improved) 
  - change the sending hook at the end of the main():
      do_classical_hook() vs. do_random_hook()

        
Tue May 28 17:56:28 CEST 2002                               0.0.10

  [libsk / arp-sk]
  - arp-sk.c: add option --rand-time to randomize the period between the
    sending of 2 packets
  - send-packets.c: 
      - add support for a random period of sending
      - add timestamp to the display

  [libsk]
  - add 2 new functions in sk_arp_pkt_handler (get_rand_hwa() and
    get_rand_ip()), to allow the use of the "rand" flags.
        
  [arp-sk]
  - control that only one -d (resp. -D, -s, -S) is used only once.
  - who-has packet dst MAC is set to 00:00:00:00:00:00 if nothing is
    specified with -D

Tue May 21 09:00:06 CEST 2002                               0.0.9

  [arp-sk]
  Very nive help output now

  [libsk / arp-sk]
  - Change str2ethip() to support :MAC only input.
  - Add show_version()

  [libsk]
  send-packets.c: segfault occurs when packets have to be send very
  quickly and that ctrl+c is entered. It is cause in cleanup() because:
    - it calls print_stats() which calls libnet_name2addr() (and not
      libnet_name2addr_r() which is buggy)
    - it attempts to write to stdout, which mess up completely with the
      writings done in send_packets().
  It can be any of the 2 reasons ... or another one, I can't say :-(
  This is currently fixed by blocking some signals in cleanup before
  exiting. 
        
Fri May 17 13:49:32 CEST 2002                               0.0.8
        
  [arp-sk]
  -s and -d both accept IP or MAC addresses.

src/ip-common.c
/**
 *
 *     CUT AND PASTE FROM nmap-3.0
 *     All honors to Fyodor <fyodor@insecure.org>
 *
 * Returns 1 if this is a reserved IP address, where "reserved" means
 * either a private address, non-routable address, or even a non-reserved
 * but unassigned address which has an extremely high probability of being
 * black-holed.
 *
 * We try to optimize speed when ordering the tests. This optimization
 * assumes that all byte values are equally likely in the input.
 *
 * Warning: This function could easily become outdated if the IANA
 * starts to assign some more IPv4 ranges to RIPE, etc. as they have
 * started doing this year (2001), for example 80.0.0.0/4 used to be
 * completely unassigned until they gave 80.0.0.0/7 to RIPE in April
 * 2001 (www.junk.org is an example of a new address in this range).
 *
 * Check <http://www.iana.org/assignments/ipv4-address-space> for
 * the most recent assigments.
 */

#include "sk.h"


int
ip_is_reserved(struct in_addr *ip)
{

    char *ipc = (char *)&(ip->s_addr);
    unsigned char i1 = ipc[0], i2 = ipc[1], i3 = ipc[2], i4 = ipc[3];

    /*
     * 221-223/8 is IANA reserved 
     * 224-239/8 is all multicast stuff 
     * 240-255/8 is IANA reserved 
     */
    if(i1 >= 221)
	return 1;

    /*
     * 096-126/8 is IANA reserved 
     * 127/8 is reserved for loopback 
     */
    if(i1 >= 96 && i1 <= 127)
	return 1;

    /*
     * 069-079/8 is IANA reserved 
     */
    if(i1 >= 69 && i1 <= 79)
	return 1;

    /*
     * 082-095/8 is IANA reserved 
     */
    if(i1 >= 82 && i1 <= 95)
	return 1;

    /*
     * do all the /7's and /8's with a big switch statement, hopefully the
     * compiler will be able to optimize this a little better using a jump 
     * table or what have you
     */
    switch (i1)
    {
	case 0:		/* 000/8 is IANA reserved       */
	case 1:		/* 001/8 is IANA reserved       */
	case 2:		/* 002/8 is IANA reserved       */
	case 5:		/* 005/8 is IANA reserved       */
	case 6:		/* USA Army ISC                 */
	case 7:		/* used for BGP protocol        */
	case 10:		/* the infamous 10.0.0.0/8      */
	case 23:		/* 023/8 is IANA reserved       */
	case 27:		/* 027/8 is IANA reserved       */
	case 31:		/* 031/8 is IANA reserved       */
	case 36:		/* 036/8 is IANA reserved       */
	case 37:		/* 037/8 is IANA reserved       */
	case 39:		/* 039/8 is IANA reserved       */
	case 41:		/* 041/8 is IANA reserved       */
	case 42:		/* 042/8 is IANA reserved       */
	case 55:		/* misc. U.S.A. Armed forces    */
	case 58:		/* 058/8 is IANA reserved       */
	case 59:		/* 059/8 is IANA reserved       */
	case 60:		/* 060/8 is IANA reserved       */
	case 197:
	    return 1;
	default:
	    break;
    }

    /*
     * 172.16.0.0/12 is reserved for private nets by RFC1819 
     */
    if(i1 == 172 && i2 >= 16 && i2 <= 31)
	return 1;

    /*
     * 192.168.0.0/16 is reserved for private nets by RFC1819 
     * 192.0.2.0/24 is reserved for documentation and examples 
     */
    if(i1 == 192)
    {
	if(i2 == 168)
	    return 1;
	else if(i2 == 0 && i3 == 2)
	    return 1;
    }

    /*
     * reserved for DHCP clients seeking addresses, not routable outside LAN 
     */
    if(i1 == 169 && i2 == 254)
	return 1;

    /*
     * believe it or not, 204.152.64.0/23 is some bizarre Sun proprietary
     * clustering thing 
     */
    if(i1 == 204 && i2 == 152 && (i3 == 64 || i3 == 65))
	return 1;

    /*
     * 255.255.255.255, note we already tested for i1 in this range 
     */
    if(i2 == 255 && i3 == 255 && i4 == 255)
	return 1;

    return 0;

}

void
get_rand_ip(u_long * ip)
{

    struct in_addr _ip;

    do
    {
	_ip.s_addr = random();
    }
    while(ip_is_reserved(&_ip));

    *ip = _ip.s_addr;
}
